src/server/dragonfly_test.cc

```diff
@@ -749,6 +756,76 @@ TEST_F(DflyEngineTest, Bug496) {
   });
 }
 
+TEST_F(DefragDflyEngineTest, TestDefragOption) {
+  // Fill data into dragonfly and then check if we have
+  // any location in memory to defrag. See issue #448 for details about this.
+  EXPECT_GE(max_memory_limit, 100'000);
+  const int NUMBER_OF_KEYS = 184000;
+
+  std::vector<std::string> keys2delete;
+  keys2delete.push_back("del");
+  for (int i = 1; i < 10; i++) {
+    keys2delete.push_back("key-name:" + std::to_string(i + 9));
+  }
+  for (int i = 10; i < 100; i++) {
+    keys2delete.push_back("key-name:" + std::to_string(i + 99));
+  }
+  for (int i = 100; i < 1000; i++) {
+    keys2delete.push_back("key-name:" + std::to_string(i + 999));
+  }
+  for (int i = 1000; i < 10000; i++) {
+    keys2delete.push_back("key-name:" + std::to_string(i + 9999));
+  }
+
+  std::vector<std::string_view> keys(keys2delete.begin(), keys2delete.end());
+
+  RespExpr resp = Run({"DEBUG", "POPULATE", std::to_string(NUMBER_OF_KEYS), "key-name", "130"});
+  ASSERT_EQ(resp, "OK");
+  resp = Run({"DBSIZE"});
+  EXPECT_THAT(resp, IntArg(NUMBER_OF_KEYS));
+
+  shard_set->pool()->AwaitFiberOnAll([&](unsigned index, ProactorBase* base) {
+    EngineShard* shard = EngineShard::tlocal();
+    ASSERT_FALSE(shard == nullptr);  // we only have one and its should not be empty!
+    EXPECT_EQ(shard->GetDefragStats().success_count, 0);
+    // we are expecting to have at least one try by now
+    EXPECT_GT(shard->GetDefragStats().tries, 0);
+    EXPECT_GT(GetMallocCurrentCommitted(), NUMBER_OF_KEYS);
+  });
+
+  ArgSlice delete_cmd(keys);
+  auto r = CheckedInt(delete_cmd);
+  ASSERT_EQ(r, keys2delete.size() - 1);
+
+  shard_set->pool()->AwaitFiberOnAll([&](unsigned index, ProactorBase* base) {
+    EngineShard* shard = EngineShard::tlocal();
+    ASSERT_FALSE(shard == nullptr);  // we only have one and its should not be empty!
+    int tries = 10;
+    bool success = base->AwaitBrief([&]() mutable {
+      // we need to await for the condition for running the
+      // on idle task to run, since there is no way to know
+      // in advance when this will happen
+      // we will run a task here that will let the idle task
+      // a chance to run. Since there is no condition to wait on
+      // this is required in order to somehow being able to
+      // let a chance for the scan to run successfully
+      auto stats = shard->GetDefragStats();
+      auto start_tries = stats.tries;
+      while (tries > 0 && stats.tries < 3) {
+        stats = shard->GetDefragStats();
+        if (stats.tries > start_tries || start_tries > 1) {
+          EXPECT_GT(stats.success_count, 3);
```

Comment:
Why this low restriction? Shouldn't we check that memory utilization are "okay" now

By the way, its hard to tell from its name what success_count means, maybe call it reallocated_count or smth similar?

Reply:
There is no direct way to tell what the memory success criteria is. I think that the most important thing is to make sure that this is stable. But you have a point here. We may need to make sure that this is not running again (or rather it does nothing). The problem is, even now it take about 0.5 second to run. bottom line I will take a look at this.